lib/model/model.go

```diff
@@ -1316,19 +1319,43 @@ func (m *Model) closeLocked(device protocol.DeviceID) {
 	closeRawConn(conn)
 }
 
-// Request returns the specified data segment by reading it from local disk.
-// Implements the protocol.Model interface.
-func (m *Model) Request(deviceID protocol.DeviceID, folder, name string, offset int64, hash []byte, weakHash uint32, fromTemporary bool, buf []byte) error {
-	if offset < 0 {
-		return protocol.ErrInvalid
+// Implements protocol.RequestResponse
+type requestResponse struct {
+	data   []byte
+	closed chan struct{}
+}
+
+func newRequestResponse(size int) *requestResponse {
+	return &requestResponse{
+		data:   protocol.BufferPool.Get(size),
+		closed: make(chan struct{}),
 	}
+}
 
-	if cfg, ok := m.cfg.Folder(folder); !ok || !cfg.SharedWith(deviceID) {
-		l.Warnf("Request from %s for file %s in unshared folder %q", deviceID, name, folder)
-		return protocol.ErrNoSuchFile
-	} else if cfg.Paused {
-		l.Debugf("Request from %s for file %s in paused folder %q", deviceID, name, folder)
-		return protocol.ErrInvalid
+func (r *requestResponse) Data() []byte {
+	return r.data
+}
+
+// Returns the byte slice back to the pool and releases the bytes to the limiter.
+func (r *requestResponse) Close() {
+	select {
+	case <-r.closed:
+		return
+	default:
+	}
+	protocol.BufferPool.Put(r.data)
+	close(r.closed)
```

Comment:
I'd move this in the default case, as this can be racy.

Reply:
Good catch...  
How will moving it into the default clause make it not racy though?

Reply:
it would not make it not racy I guess, just makes it less likely